pkg/api/vminfo.go

```diff
@@ -337,29 +337,77 @@ func (b ResourceBounds) validate(ec *erc.Collector, path string, memSlotSize *re
 // ScalingConfig provides bits of configuration for how the autoscaler-agent makes scaling decisions
 type ScalingConfig struct {
 	// LoadAverageFractionTarget sets the desired fraction of current CPU that the load average
-	// should be. For example, with a value of 0.7, we'd want load average to sit at 0.7 ×
-	// CPU,
+	// should be. For example, with a value of 0.7, we'd want load average to sit at 0.7 × CPU,
 	// scaling CPU to make this happen.
-	LoadAverageFractionTarget float64 `json:"loadAverageFractionTarget"`
+	//
+	// When specifying the autoscaler-agent config, this field is required. For an individual VM, if
+	// this field is left out the settings will fall back on the global default.
+	LoadAverageFractionTarget *float64 `json:"loadAverageFractionTarget,omitempty"`
 
 	// MemoryUsageFractionTarget sets the desired fraction of current memory that
 	// we would like to be using. For example, with a value of 0.7, on a 4GB VM
 	// we'd like to be using 2.8GB of memory.
-	MemoryUsageFractionTarget float64 `json:"memoryUsageFractionTarget"`
+	//
+	// When specifying the autoscaler-agent config, this field is required. For an individual VM, if
+	// this field is left out the settings will fall back on the global default.
+	MemoryUsageFractionTarget *float64 `json:"memoryUsageFractionTarget,omitempty"`
 }
 
-func (c *ScalingConfig) Validate() error {
+// WithOverrides returns a new copy of defaults, where fields set in overrides replace the ones in
+// defaults but all others remain the same.
+//
+// overrides may be nil; if so, this method just returns defaults.
+func (defaults ScalingConfig) WithOverrides(overrides *ScalingConfig) ScalingConfig {
+	if overrides == nil {
+		return defaults
+	}
+
+	if overrides.LoadAverageFractionTarget != nil {
+		defaults.LoadAverageFractionTarget = &[]float64{*overrides.LoadAverageFractionTarget}[0] // slice trick to shallow copy
```

Comment:
Did you mean deep copy?

Also, WDYT about https://go.dev/play/p/wDUnmOqCSqC ?

Reply:
Hm... control plane uses https://github.com/samber/lo's [`lo.ToPtr`](https://pkg.go.dev/github.com/samber/lo#ToPtr) to fulfill this purpose - see e.g. here:

https://github.com/neondatabase/cloud/blob/2b0562e85d4586ad62a8c7af84fbc0af647391f0/goapp/internal/operationsv2/provisioner/provisioner_k8s_common.go#L289-L293

We also define a similar helper function in `pkg/agent/core/state_test.go`:

https://github.com/neondatabase/autoscaling/blob/4d83eb6efe45b08ea8d8a1c93d69013a1a23b549/pkg/agent/core/state_test.go#L234-L236

I'm hesitant to introduce a temporary function for just this location, but I'd be happy to open a follow-up PR to tackle this repo-wide ?

WDYT?

Reply:
> en a follow-up PR to tackle this repo-wide ?

Sure, when you have the time.

Reply:
PR here: #946